fix(frappe.utils.data): Deprecate `cast_fieldtype` to use `cast` for consistent return types

gavindsouza opened the pull request:
Bug noticed via the change https://github.com/frappe/frappe/pull/13655 made in `frappe.db.get_single_value`. cast_fieldtype would return current date for Date fields, current datetime for Datetime fields and None for Time field if value was `None`.

### Why did this bug happen?

![Screenshot 2021-08-20 at 11 25 40 AM](https://user-images.githubusercontent.com/36654812/130186640-ef8e58d5-c9e5-4abf-af36-71ed2497a482.png)

### How should this be resolved?

Probably by `cast_fieldtype` returning Date and Datetime for None input as: ` datetime.date(0001, 1, 1)`, `datetime.datetime(0001, 1, 1, 0, 0)`...But unsure about that. First solution to this was _"Hey, maybe let's just not cast `Falsy`values?"_ (first commit in this branch), but the deeper problem with cast_fieldtype.

---

So how's this? Does the new change fix this issue? Probably (from my POV...if there's still an issue, drop a comment).

![Screenshot 2021-08-20 at 11 56 35 AM](https://user-images.githubusercontent.com/36654812/130191583-e526dedb-99ad-453f-9584-ebcfcc4863a9.png)
 
---

But, I don't want to break this API, maybe the "right"_(-er)_ way would be to make a new util, perhaps `frappe.utils.data.cast` instead and deprecate `cast_fieldtype`?

Okay, so I did that in the third and fourth commits. Added the new util and replaced the usage everywhere else. Reverted `frappe.utils.data.cast_fieldtype` to it's behaviour before this PR (except for that stdout warning).

Added a deprecation warning to stdout but also want to raise a `DeprecationWarning`...realised we don't have a good standard way of doing this yet...Thinking of adding https://github.com/briancurtin/deprecation for that.

---

ToDo:

- [x] Add deprecation warning: Added conditionally (if not used from `BaseDocument.cast`)
- [x] Add tests

---

Ref: https://github.com/frappe/frappe/pull/13655#issuecomment-902045672

gavindsouza committed:
fix(db): Cast single dt field only if value is truthy

gavindsouza committed:
fix: Consistent return types in cast_fieldtype

Note: BREAKING CHANGE

gavindsouza committed:
feat: frappe.utils.data.cast

Cast the value to the Python native object of the Frappe fieldtype provided. If value is None, the first/lowest value of the `fieldtype` will be returned.

Mapping of Python types => Frappe types:
* float => ("Currency", "Float", "Percent")
* int => ("Int", "Check")
* datetime.datetime => ("Datetime",)
* datetime.date => ("Date",)
* datetime.time => ("Time",)

Deprecate frappe.utils.data.cast_fieldtype in favour of new util cast
which handles types "better"

gavindsouza committed:
fix: Use cast in favour of cast_fieldtype

Use newly introduced casting util for Python-Frappe types mapping

surajshetty3416 commented:
LGTM

ankush commented on frappe/model/base_document.py:
```diff
@@ -969,7 +969,7 @@ def get_value(self, fieldname):
 
 	def cast(self, value, df):
-		return cast_fieldtype(df.fieldtype, value)
+		return cast(df.fieldtype, value)
```
Other than this change... LGTM.

The previous behaviour of cast_fieldtype (`None => current_time`) might have been used while writing business logic sadly. So this becomes a breaking change and could introduce bugs 😅  

On the flip side, expecting `None` to evaluate to current time instead of just using `now()`, `nowdate()` etc is just as bad. But ... 

https://github.com/frappe/frappe/blob/1ac27efa7a252701e4c7c61c8aacc90fe401cb99/frappe/utils/data.py#L49-L53

🤷

rmehta replied:
I agree with @ankush

This is a breaking change - might cause a lot of downstream problems.

gavindsouza committed:
chore: Update frappe.utils.data.cast docstring

gavindsouza committed:
fix: Revert to using cast_fieldtype in BaseDocument.cast

* reference: revert Breaking Change -
  https://github.com/frappe/frappe/pull/13989#discussion_r695624003
* Show deprecation warning unless `show_warning` is unset

gavindsouza committed:
Merge branch 'develop' into get-single-value-oof

gavindsouza committed:
test: Add tests for frappe.utils.data.cast

gavindsouza committed:
test: Add tests for str fieldstypes cast

gavindsouza committed:
feat(utils): Add util get_timedelta

get_timedelta returns None in case of invalid or imparsable inputs. This
behaviour is consistent wrt other utils. The util, to_timedelta tries to
convert to timedelta objects only if str object is passed. It returns
the same object if not string, which is absurd...given its called
`to_timedelta`.

gavindsouza committed:
fix: Use get_timedelta in cast, add to safe_exec list

* Return date and datetime objects for Date and Datetime field types
  respectively if Falsy output is set

gavindsouza committed:
fix: Check if input is str is_invalid_date_string

This is sort of a breaking change? Because if an int/dict/list/tuple was
passed instead of a str, object doesnt have .startswith (AttributeError)
would be raised instead of just returning None.

ankush commented:
LGTM :+1:

ankush commented on frappe/utils/data.py:
```diff
@@ -16,10 +18,10 @@
 
 # datetime functions
-def getdate(string_date=None):
+def getdate(string_date: str = None):
```
```suggestion
def getdate(string_date: Optional[str] = None):
```

Already imported.

ankush commented on frappe/utils/data.py:
```diff
@@ -67,6 +69,31 @@ def get_datetime(datetime_str=None):
 	except ValueError:
 		return parser.parse(datetime_str)
 
+def get_timedelta(time: str = None) -> Optional[datetime.timedelta]:
```
```suggestion
def get_timedelta(time: Optional[str] = None) -> Optional[datetime.timedelta]:
```

ankush reviewed (commented):
minor type hint changes.

gavindsouza committed:
chore(utils): Add type hints for get_timedelta, getdate

Co-authored-by: Ankush Menat <ankushmenat@gmail.com>

gavindsouza commented:
@Mergifyio backport version-13-hotfix version-12-hotfix

gavindsouza commented:
Realised external docs may not be required since this is an internal API, and all the relevant info is already documented via code docstrings, comments and tests
